crazy_rl/multi_agent/jax/base_parallel_env.py

```diff
@@ -0,0 +1,389 @@
+            self.drone_ids = [i for i in range(num_drones)]
+            assert swarm is not None, "Swarm object must be provided in real mode"
+            self.swarm = swarm
+            while not self.swarm:
+                time.sleep(0.5)
+                print("Waiting for connection...")
+        """
+
+    def _observation_space(self, agent) -> spaces.Space:
+        """Returns the observation space of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _action_space(self) -> spaces.Space:
+        """Returns the action space of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_obs(self, state):
+        """Returns the current observation of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_action(self, actions, state):
+        """Computes the action passed to `.step()` into action matching the mode environment. Must be implemented in a subclass.
+
+        Args:
+            actions : ndarray | dict[..]. The input action for one drones
+            state : the state of the environment (contains agent_location used in this function)
+        """
+        raise NotImplementedError
+
+    def _compute_reward(self, state):
+        """Computes the current reward value(s). Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_terminated(self, state):
+        """Computes the current done value(s). Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_truncation(self, state):
+        """Computes the current done value(s). Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    # PettingZoo API
+    @override
+    def reset(self, state, seed=None, return_info=False, options=None):
+        """
+        if self._mode == "real":
+            # self.swarm.parallel_safe(reset_estimator)
+            state = jdc.replace(state, agent_location=self._get_drones_state(self._mode, state.agent_location))
+            print("reset", state.agent_location)
+
+            command = dict()
+            # dict target_position URI
+            for agent in range(self.num_drones):
+                uri = "radio://0/4/2M/E7E7E7E7" + str(agent).zfill(2)
+                target = self._init_flying_pos[agent]
+                agent = state.agent_location[agent]
+                command[uri] = [[agent, target]]
+
+            self.swarm.parallel_safe(run_take_off)
+            print("Take off successful.")
+            print(f"Setting the drone positions to the initial positions. {command}")
+            self.swarm.parallel_safe(run_sequence, args_dict=command)
+
+            state = jdc.replace(state, agent_location=self._get_drones_state(self._mode, state.agent_location))
+
+        else:
+        """
+        state = jdc.replace(state, agent_location=jnp.copy(self._init_flying_pos))
```

Comment:
Why not just recreating a state from the start? To me, reset should not take a state as argument since it does not care about the current state.